### PR TITLE
refactor: GRC20

### PR DESCRIPTION
**Pull request by leohhhn**
## Description

Closes: #2294

This PR refactors the GRC20 folder to follow the ERC20 spec as close as possibly.

<details><summary>Contributors' checklist...</summary>

- [x] Added new tests, or not needed, or not feasible
- [x] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [x] Updated the official documentation or not needed
- [x] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [x] Added references to related issues and PRs
- [ ] Provided any useful hints for running manual tests
- [ ] Added new benchmarks to [generated graphs](https://gnoland.github.io/benchmarks), if any. More info [here](https://github.com/gnolang/gno/blob/master/.benchmarks/README.md).
</details>

### CODE REVIEW

**Comment by leohhhn** [comment]
Interesting discussion point: https://github.com/gnolang/gno/issues/2316

**Comment by leohhhn** [comment]
Opening up for reviews - grc20 package, foo20, wugnot & grc20 factory realms have been refactored, need to fix up some other stuff that is failing due to the changed API.

**Comment by zivkovicmilos** [comment]
Please fix the CI before opening the PR 🙏

**Review by tbruyelle — APPROVED** [review]
I really like the simplicity of this version.

With `grc20factory` that acts as a grc20 multiplexer, I wonder if we still need to have the `IGRC20` interface. This interface was initially designed to make `*grc20.Token` implementation shareable between realms, but now these realms can just use the `grc20factory` right ?

**Comment by leohhhn** [comment]
@tbruyelle I think both have valid use cases. People can choose to use the factory, or simply write their own realm importing the GRC20 package.

File: examples/gno.land/p/demo/grc/grc20/grc20.gno [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+// Make sure Token implements IGRC20 interface
+var _ IGRC20 = (*Token)(nil)
+
+const (
+	emptyAddress  = std.Address("")
+	TransferEvent = "Transfer"
+	ApprovalEvent = "Approval"
+)
+
+func NewGRC20Token(name, symbol string, decimals uint8) *Token {
```
Comment (moul) [inline examples/gno.land/p/demo/grc/grc20/grc20.gno:39]:
```suggestion
func NewToken(name, symbol string, decimals uint8) *Token {
```

grc20 is the package name.

Comment (moul) [reply examples/gno.land/p/demo/grc/grc20/grc20.gno:39]: i think we should not have a `NewToken` helper, but only a `NewBanker` one.

File: examples/gno.land/p/demo/grc/grc20/grc20.gno [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+	"math"
+	"math/bits"
+	"std"
+	"strconv"
+
+	"gno.land/p/demo/avl"
+	"gno.land/p/demo/ufmt"
+)
+
+type Token struct {
```
Comment (moul) [inline examples/gno.land/p/demo/grc/grc20/grc20.gno:21]: This looks good, but you need to have a `type Banker interface { Token /* embedded */, Mint() }` which is what you initializes first.

File: examples/gno.land/p/demo/grc/grc20/grc20.gno [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+func (t *Token) Decimals() (decimals uint8) {
+	return t.decimals
+}
+
+func (t *Token) TotalSupply() (totalSupply uint64) {
+	return t.totalSupply
+}
+
+func (t *Token) BalanceOf(owner std.Address) (balance uint64) {
+	mustBeValid(owner)
```
Comment (moul) [inline examples/gno.land/p/demo/grc/grc20/grc20.gno:66]: we can remove this one.

File: examples/gno.land/p/demo/grc/grc20/grc20.gno [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+func (t *Token) Allowance(owner, spender std.Address) (remaining uint64) {
+	rawAllowance, found := t.allowances.Get(spenderKey(owner, spender))
+	if !found {
+		return 0
+	}
+
+	return rawAllowance.(uint64)
+}
+
+func (t *Token) Mint(account std.Address, amount uint64) {
```
Comment (moul) [inline examples/gno.land/p/demo/grc/grc20/grc20.gno:107]:
```suggestion
func (t *Banker) Mint(account std.Address, amount uint64) {
```

File: examples/gno.land/p/demo/grc/grc20/grc20.gno (hunk 2 above)
Comment (moul) [reply examples/gno.land/p/demo/grc/grc20/grc20.gno:21]: the banker should also have all the "admin" helpers, such as ways to call Transfer() and specifying a custom `from`.

File: examples/gno.land/p/demo/grc/grc20/grc20.gno [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+func (t *Token) update(from, to std.Address, value uint64) {
+	// If new tokens are minted, check for overflow
+	if from == emptyAddress {
+		sum, carry := bits.Add64(t.totalSupply, value, 0)
+		if carry != 0 {
+			err := ufmt.Sprintf("GRC20($%s): uint64 overflow: %d + %d", t.symbol, t.totalSupply, value)
+			panic(err)
+		}
+
+		t.totalSupply = sum
```
Comment (moul) [inline examples/gno.land/p/demo/grc/grc20/grc20.gno:127]: shouldn't you also mint the tokens to the `to` address?

File: examples/gno.land/r/demo/foo20/foo20.gno [hunk 6]
```diff
@@ -1,112 +1,143 @@
...
+	fauceted bool
 )
 
 func init() {
-	foo = grc20.NewAdminToken("Foo", "FOO", 4)
-	foo.Mint(admin, 1000000*10000)                                    // @administrator (1M)
-	foo.Mint("g1u7y667z64x2h7vc6fmpcprgey4ck233jaww9zq", 10000*10000) // @manfred (10k)
+	// Instantiate ownable
+	o = ownable.New()
+	o.TransferOwnership("g125em6arxsnj49vx35f0n0z34putv5ty3376fg5") // @admin
```
Comment (moul) [inline examples/gno.land/r/demo/foo20/foo20.gno:25]: Let's remove the token owner completely and expose a "Faucet" function for the demonstration.

See #2388 for an example.

File: examples/gno.land/r/demo/wugnot/z0_filetest.gno
Comment (moul) [inline examples/gno.land/r/demo/wugnot/z0_filetest.gno:1]: I prefer if you keep this file as a test, or else, please implement it completely using standard unit tests.

File: examples/gno.land/r/demo/wugnot/wugnot.gno [hunk 7]
```diff
@@ -82,49 +80,51 @@ func Render(path string) string {
...
 
+// Approve allows `spender` to withdraw from your account multiple times, up to the `value` amount
+// If this function is called again it overwrites the current allowance with value
 func Approve(spender pusers.AddressOrName, amount uint64) {
-	caller := std.PrevRealm().Addr()
-	err := wugnot.Approve(caller, users.Resolve(spender), amount)
-	if err != nil {
-		panic(err)
-	}
+	wugnot.Approve(users.Resolve(spender), amount)
```
Comment (moul) [inline examples/gno.land/r/demo/wugnot/wugnot.gno:124]: why removing the panic?

File: examples/gno.land/r/demo/wugnot/wugnot.gno [hunk 8]
```diff
@@ -68,10 +67,9 @@ func Render(path string) string {
 	switch {
 	case path == "":
 		return wugnot.RenderHome()
-	case c == 2 && parts[0] == "balance":
-		owner := std.Address(parts[1])
-		balance, _ := wugnot.BalanceOf(owner)
-		return ufmt.Sprintf("%d\n", balance)
+	case c == 2 && parts[0] == "balance": // pkgpath:balance/address_or_username
+		owner := pusers.AddressOrName(parts[1])
+		return ufmt.Sprintf("%d\n", wugnot.BalanceOf(users.Resolve(owner)))
```
Comment (moul) [inline examples/gno.land/r/demo/wugnot/wugnot.gno:72]:
```suggestion
		balance := wugnot.BalanceOf(users.Resolve(owner))
		return ufmt.Sprintf("%d\n", balance)
```

File: examples/gno.land/p/demo/grc/grc20/grc20.gno [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+}
+
+// spendAllowance updates `owner`s allowance for `spender` based on spent `value`
+// Does not update the allowance value in case of infinite allowance
+// Panics if not enough allowance is available
+// Does not emit an Approval event
+func (t *Token) spendAllowance(owner, spender std.Address, value uint64) {
+	currentAllowance := t.Allowance(owner, spender)
+
+	if currentAllowance != math.MaxUint64 {
```
Comment (moul) [inline examples/gno.land/p/demo/grc/grc20/grc20.gno:199]: This exception should be documented. If I'm not mistaken, you consider a MaxUint64 allowance to be equivalent to an infinite allowance, correct?

File: examples/gno.land/r/demo/grc20factory/grc20factory_test.gno [hunk 10]
```diff
@@ -1,17 +1,21 @@
...
+	"gno.land/p/demo/testutils"
 )
 
 func TestReadOnlyPublicMethods(t *testing.T) {
-	admin := std.Address("g1us8428u2a5satrlxzagqqa5m6vmuze025anjlj")
-	manfred := std.Address("g1u7y667z64x2h7vc6fmpcprgey4ck233jaww9zq")
-	unknown := std.Address("g1jg8mtutu9khhfwc4nxmuhcpftf0pajdhfvsqf5") // valid but never used.
+
+	admin := testutils.TestAddress("admin")     // g1v9kxjcm9ta047h6lta047h6lta047h6lzd40gh
+	manfred := testutils.TestAddress("manfred") // g1wymu47drhr0kuq2098m792lytgtj2nyx77yrsm
```
Comment (moul) [inline examples/gno.land/r/demo/grc20factory/grc20factory_test.gno:13]:
```suggestion
	alice := testutils.TestAddress("alice") // xxx
```

let's keep "manfred" when it's really my address.

File: examples/gno.land/r/demo/grc20factory/grc20factory.gno [hunk 11]
```diff
@@ -1,140 +1,185 @@
...
+	newToken := grc20.NewGRC20Token(name, symbol, decimals)
 	newToken.Mint(admin, initialMint)
-	t := token{
-		adminToken: newToken,
-		admin:      admin,
+
+	t := tokenWithAdmin{
+		token:    newToken,
+		admin:    admin,
+		fauceted: false, // faucet disabled by default
```
Comment (moul) [inline examples/gno.land/r/demo/grc20factory/grc20factory.gno:43]: If disabled by default, please register one or two tokens in `init()` that will be fauceted as true by default, so this package can be easily tested.

**Comment by moul** [comment]
#2529

**Comment by leohhhn** [comment]
Clsoing this in favor of @moul's refactor.

**Pull request closed without merging**